File: Backend/src/repositories/invoiceRepository.ts
import { getPool, withTransaction } from '../db';
import type { Invoice, InvoiceItem } from '../types';

const COLUMNS = 'id, invoice_number, items, subtotal, discount, tax_percent, tax_amount, total, notes, created_at';
const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const map = (row: any): Invoice => ({
  id: row.id,
  invoiceNumber: row.invoice_number,
  items: Array.isArray(row.items) ? row.items : [],
  subtotal: row.subtotal,
  discount: row.discount,
  taxPercent: row.tax_percent,
  taxAmount: row.tax_amount,
  total: row.total,
  notes: row.notes,
  createdAt: row.created_at,
});

export type NewInvoice = Omit<Invoice, 'id' | 'invoiceNumber' | 'createdAt'>;

export const invoiceRepository = {
  async getAll(): Promise<Invoice[]> {
    const { rows } = await getPool().query(`SELECT ${COLUMNS} FROM invoices ORDER BY created_at DESC LIMIT 1000`);
    return rows.map(map);
  },

  async getById(id: string): Promise<Invoice | null> {
    const { rows } = await getPool().query(`SELECT ${COLUMNS} FROM invoices WHERE id = $1`, [id]);
    return rows[0] ? map(rows[0]) : null;
  },

  async remove(id: string): Promise<boolean> {
    const result = await getPool().query('DELETE FROM invoices WHERE id = $1', [id]);
    return (result.rowCount ?? 0) > 0;
  },

  /**
   * One transaction: bump the global counter, insert the invoice, and (when inventory tracking is on) decrement
   * stock for each line whose productId is a UUID. No availability check - stock may go negative, as before.
   */
  async create(invoice: NewInvoice, decrementStock: boolean): Promise<Invoice> {
    return withTransaction(async (client) => {
      const seqResult = await client.query("UPDATE counters SET seq = seq + 1 WHERE id = 'invoice' RETURNING seq");
      const seq: number = seqResult.rows[0].seq;

      const now = new Date();
      const yyyymm = `${now.getUTCFullYear()}${String(now.getUTCMonth() + 1).padStart(2, '0')}`;
      const invoiceNumber = `VD-${yyyymm}-${String(seq).padStart(4, '0')}`;

      const inserted = await client.query(
        `INSERT INTO invoices (invoice_number, items, subtotal, discount, tax_percent, tax_amount, total, notes)
         VALUES ($1, $2::jsonb, $3, $4, $5, $6, $7, $8)
         RETURNING ${COLUMNS}`,
        [
          invoiceNumber,
          JSON.stringify(invoice.items),
          invoice.subtotal,
          invoice.discount,
          invoice.taxPercent,
          invoice.taxAmount,
          invoice.total,
          invoice.notes,
        ],
      );
      const created = map(inserted.rows[0]);

      if (decrementStock) {
        for (const item of invoice.items as InvoiceItem[]) {
          if (UUID_RE.test(item.productId)) {
            await client.query('UPDATE products SET stock = stock - $1 WHERE id = $2', [item.quantity, item.productId]);
          }
        }
      }

      return created;
    });
  },
};

File: Backend/VijayDairy.Postgres/InvoiceRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Text.Json;
using System.Threading.Tasks;
using VijayDairy.Domain.Entities;
using VijayDairy.Persistence;

namespace VijayDairy.Postgres;

public class InvoiceRepository : PostgresHelper, IInvoiceRepository
{
    private static readonly JsonSerializerOptions JsonOpts = new(JsonSerializerDefaults.Web);

    public InvoiceRepository(string connectionString) : base(connectionString) { }

    private static Invoice Map(IDataReader reader) => new()
    {
        Id = reader.GetGuid(reader.GetOrdinal("id")),
        InvoiceNumber = reader.GetString(reader.GetOrdinal("invoice_number")),
        Items = JsonSerializer.Deserialize<List<InvoiceItem>>(reader.GetString(reader.GetOrdinal("items")), JsonOpts) ?? new(),
        Subtotal = reader.GetDecimal(reader.GetOrdinal("subtotal")),
        Discount = reader.GetDecimal(reader.GetOrdinal("discount")),
        TaxPercent = reader.GetDecimal(reader.GetOrdinal("tax_percent")),
        TaxAmount = reader.GetDecimal(reader.GetOrdinal("tax_amount")),
        Total = reader.GetDecimal(reader.GetOrdinal("total")),
        Notes = reader.GetString(reader.GetOrdinal("notes")),
        CreatedAt = reader.GetDateTime(reader.GetOrdinal("created_at")),
    };

    private const string SelectColumns =
        "id, invoice_number, items::text AS items, subtotal, discount, tax_percent, tax_amount, total, notes, created_at";

    public Task<List<Invoice>> GetAllAsync()
    {
        string sql = $"SELECT {SelectColumns} FROM invoices ORDER BY created_at DESC LIMIT 1000";
        using var reader = ExecuteDataReader(sql);
        var results = new List<Invoice>();
        while (reader.Read()) results.Add(Map(reader));
        return Task.FromResult(results);
    }

    public Task<Invoice?> GetByIdAsync(Guid id)
    {
        string sql = $"SELECT {SelectColumns} FROM invoices WHERE id = @Id";
        using var reader = ExecuteDataReader(sql, new[] { Param("@Id", id) });
        return Task.FromResult(reader.Read() ? Map(reader) : null);
    }

    public Task<bool> DeleteAsync(Guid id)
    {
        var rows = ExecuteNonQuery("DELETE FROM invoices WHERE id = @Id", new[] { Param("@Id", id) });
        return Task.FromResult(rows > 0);
    }

    public async Task<Invoice> CreateAsync(Invoice invoice, bool decrementStock)
    {
        using var connection = CreateAndOpenConnection();
        using var transaction = await connection.BeginTransactionAsync();
        try
        {
            using var seqCommand = CreateCommand(connection, "UPDATE counters SET seq = seq + 1 WHERE id = 'invoice' RETURNING seq");
            seqCommand.Transaction = transaction;
            var seq = (int)(await seqCommand.ExecuteScalarAsync())!;

            invoice.InvoiceNumber = $"VD-{DateTime.UtcNow:yyyyMM}-{seq:D4}";
            var itemsJson = JsonSerializer.Serialize(invoice.Items, JsonOpts);

            string insertSql = $@"
                INSERT INTO invoices (invoice_number, items, subtotal, discount, tax_percent, tax_amount, total, notes)
                VALUES (@InvoiceNumber, @Items::jsonb, @Subtotal, @Discount, @TaxPercent, @TaxAmount, @Total, @Notes)
                RETURNING {SelectColumns}";

            using var insertCommand = CreateCommand(connection, insertSql, new[]
            {
                Param("@InvoiceNumber", invoice.InvoiceNumber),
                Param("@Items", itemsJson),
                Param("@Subtotal", invoice.Subtotal),
                Param("@Discount", invoice.Discount),
                Param("@TaxPercent", invoice.TaxPercent),
                Param("@TaxAmount", invoice.TaxAmount),
                Param("@Total", invoice.Total),
                Param("@Notes", invoice.Notes),
            });
            insertCommand.Transaction = transaction;

            using var reader = await insertCommand.ExecuteReaderAsync();
            await reader.ReadAsync();
            var created = Map(reader);
            reader.Close();

            if (decrementStock)
            {
                foreach (var item in invoice.Items)
                {
                    if (Guid.TryParse(item.ProductId, out var productId))
                    {
                        using var stockCommand = CreateCommand(connection, "UPDATE products SET stock = stock - @Qty WHERE id = @Id", new[]
                        {
                            Param("@Qty", item.Quantity),
                            Param("@Id", productId),
                        });
                        stockCommand.Transaction = transaction;
                        await stockCommand.ExecuteNonQueryAsync();
                    }
                }
            }

            await transaction.CommitAsync();
            return created;
        }
        catch
        {
            await transaction.RollbackAsync();
            throw;
        }
    }
}
